lib/ts/recipe/accountlinking/recipe.ts

```diff
@@ -131,184 +134,208 @@ export default class Recipe extends RecipeModule {
-            }
-            logDebugMessage("createPrimaryUserIdOrLinkAccounts creating primary user");
-
-            let createPrimaryUserResult = await this.recipeInterfaceImpl.createPrimaryUser({
-                recipeUserId: user.loginMethods[0].recipeUserId,
-                userContext,
-            });
-
-            if (createPrimaryUserResult.status === "OK") {
-                logDebugMessage("createPrimaryUserIdOrLinkAccounts created primary user");
-                return createPrimaryUserResult.user;
-            }
-
-            // status is "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR" or "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR"
-
-            // if it comes here, it means that the recipe
-            // user ID is already linked to another primary
-            // user id (race condition), or that some other
-            // primary user ID exists with the same email / phone number (again, race condition).
-            // So we do recursion here to try again.
-            return await this.createPrimaryUserIdOrLinkAccounts({
+            // We try and list all users that can be linked to the input user based on the account info
+            // later we can use these when trying to link or when checking if linking to the session user is possible.
+            const linkRes = await this.tryLinkingByAccountInfo({
+                inputUser: inputUser,
+                session,
                 tenantId,
-                user,
                 userContext,
             });
-        } else {
-            logDebugMessage("createPrimaryUserIdOrLinkAccounts got linking candidate");
-            if (primaryUser.id === user.id) {
-                logDebugMessage("createPrimaryUserIdOrLinkAccounts user already linked");
-                // This can only happen cause of a race condition cause we already check
-                // if the input recipeUserId is a primary user early on in the function.
-                return user;
+            if (linkRes.status === "OK") {
+                return { status: "OK", user: linkRes.user };
             }
-            // this means that we found a primary user ID which can be linked to this recipe user ID. So we try and link them.
-
-            // we can use the 0 index cause this user is
-            // not a primary user.
-            let shouldDoAccountLinking = await this.config.shouldDoAutomaticAccountLinking(
-                user.loginMethods[0],
-                primaryUser,
-                tenantId,
-                userContext
-            );
-
-            if (!shouldDoAccountLinking.shouldAutomaticallyLink) {
-                logDebugMessage(
-                    "createPrimaryUserIdOrLinkAccounts not linking because shouldAutomaticallyLink is false"
-                );
-                return user;
+            if (linkRes.status === "NO_LINK") {
+                return { status: "OK", user: inputUser };
             }
```

Comment:
i really prefer if we don't introduce new status codes.. please reuse existing ones!